### PR TITLE
costumizer: improved file handling

### LINKED ISSUE #2307: Costumizer: replacing .scad with .json
While reading the code of the customizer, I found this:
`this->jsonFile = scadFile.replace(".scad", ".json").toStdString();`
which has the implication, that when I have a file called "this.scad" with a path like this:
`/home/userName/myProjects/files.scad/this.scad`
the json file ends up in the folder
`/home/userName/myProjects/files.json/this.json`
That is actually a cool hidden feature.

But thinking deeper, this has to be a bug/issue: Maybe someone is using a username like students.scad or mu.scadet, resulting in OpenSCAD trying to write in someone elses userprofile...

So, this has to be a bug. If no one disagrees, I would write a fix for that issue.
Most robust would be to remove the file ending from the scad file and then to add .json.
(Keeping in mind, that the file ending can be any capitalization of .sCAd, none, ...)

If someone thinks that this is a feature, we need a very clear specification to workaround issues like  usernames randomly containing .scad.

### PR DESCRIPTION
**Pull request by MichaelPFrey**
* move the writeable check from "readParameterSet" to "readFile"
   * This also fixes an issue with empty (0 bytes) json files: Empty json files cause an exception, causing readParameterSet to return false, without actually checking if the file would be writable.
* checking if the file exists, before trying to check for readability or writeabilty
*central setFile function with replace_extension instead of replace (closes #2307)

### CODE REVIEW

File: src/parameter/ParameterWidget.cc [hunk 1]
```diff
@@ -101,19 +105,37 @@ void ParameterWidget::onSetSaveButton()
...
+
+	if(exists){
+		bool readable = readParameterSet(this->jsonFile);
+
+		//check whether file is writeable
+		std::fstream file;
+		file.open(this->jsonFile, std::ios::app);
+		if (file.is_open()) {
+			file.close();
+			writeable = true;
```
Comment (kintel) [inline src/parameter/ParameterWidget.cc]: I think these 5 lines of code could be simplified to smth. like this:
```
if (std::stream{this->jsonFile, std::ios::app}) writeable = true;
```

**Comment by kintel** [comment]
Looks great! I made a suggestion to simplify some code..

File: src/parameter/ParameterWidget.cc (hunk 1 above)
Comment (MichaelPFrey) [reply src/parameter/ParameterWidget.cc]: if (std::fstream(this->jsonFile, std::ios::app)) writeable = true;

works, as in:
* it detects, when a json file is read only
* if the json file is not readonly, I can add a new parameter set and save
but I wonder how.

this is using the [constructor of fstream](http://www.cplusplus.com/reference/fstream/fstream/open/). It creates the fstream on the stack. Ah: I found the hint I needed: [RAII](https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization), the enclosing scope beeing the brackets of the "if".

Neat :-)

Comment (kintel) [reply src/parameter/ParameterWidget.cc]: I was a bit terse in my comment - glad you got it :)
Don't hesitate to ask me to clarify. Hard to gauge what detail level to add..

Comment (MichaelPFrey) [reply src/parameter/ParameterWidget.cc]: I have the hints I need, thank you. I am always a bit confused where C++ and QT are on the continuum between C and Java (for both of which I have college education). Doesn´t help that a lot of modern C++/QT text books skim over memory management and Stackoverflow tends to use C solutions for C++ problems.

Comment (kintel) [reply src/parameter/ParameterWidget.cc]: These are good reads which cover modern C++:
* A tour of C++ (https://www.amazon.com/Tour-C-Depth/dp/0321958314)
* Effective Modern C++ (e.g. https://edisciplinas.usp.br/pluginfile.php/1995323/mod_resource/content/1/Effective%20Modern%20C%2B%2B%202014.pdf)

**Comment by MichaelPFrey** [comment]
that could have coincidentally have fixed #2103
(hard to test - I was  not looking for that issue)

**Pull request merged**